highfidelity/hifi
# Case 7120, Case 14963: Procedural material entities on primitives and meshes

because you closed https://github.com/highfidelity/hifi/pull/16085

labels: engine, development ready, feedback welcome, open source, scripting API change, JSDoc

http://roadmap.highfidelity.com/open-source-project-proposals/p/procedural-materials-on-models-and-avatars
https://highfidelity.manuscript.com/f/cases/7120/Ability-for-a-user-to-create-import-custom-shaders-as-an-entity-material-type
https://highfidelity.fogbugz.com/f/cases/14963/Add-support-for-shaders-on-meshes-and-avatars

Test plan:
- A new option has been added in Settings -> Graphics and at the bottom of Developer -> Render, "Use Procedural Materials on Meshes".  It will default to disabled.
- Run [this](https://gist.githubusercontent.com/SamGondelman/df6f6c63bc8297a7d4a2dd439b72d8fb/raw/d7eb2a0e565b9e5683ab7981bca8a458436c57d3/ProceduralMaterial.js).  You'll see this:
![hifi-snap-by-SamGondelman-on-2019-05-16_20-14-34](https://user-images.githubusercontent.com/7650116/57900883-886df800-7817-11e9-8c3d-dae944bbf0f7.gif)
- When you press "u", the material entity (the small sphere) will be applied to the following in order and will then repeat: the sphere, the woody, the tablet, your avatar, nothing
- You'll see the shader on the sphere:
![hifi-snap-by-SamGondelman-on-2019-05-16_20-14-49](https://user-images.githubusercontent.com/7650116/57900896-902d9c80-7817-11e9-83fa-c1f34df607fb.gif)
- But you won't see the material on the mannequin, tablet, or yourself (it will make the targeted meshes disappear):
![hifi-snap-by-SamGondelman-on-2019-05-16_20-15-23](https://user-images.githubusercontent.com/7650116/57900906-9754aa80-7817-11e9-90a1-4775f806d308.jpg)
- Close Interface.  Add a new environment variable `HIFI_ENABLE_MESH_SHADERS` with value 1.  Restart Interface.  When you enable "Use Procedural Materials on Meshes", you will start to see the material on those things:
![hifi-snap-by-SamGondelman-on-2019-05-16_20-15-37](https://user-images.githubusercontent.com/7650116/57900917-a6d3f380-7817-11e9-97f4-f7bb71a5cceb.gif)
![hifi-snap-by-SamGondelman-on-2019-05-16_20-15-45](https://user-images.githubusercontent.com/7650116/57900927-ad626b00-7817-11e9-9f34-38e421e3337c.gif)
![hifi-snap-by-SamGondelman-on-2019-05-16_20-15-53](https://user-images.githubusercontent.com/7650116/57900931-b2271f00-7817-11e9-808e-dfca379da61d.gif)
- The setting will persist between sessions.
